### PR TITLE
[SYCL] add operator-> to annotated_arg and add e2e test

### PR DESCRIPTION
**Pull request by broxigarchen**
Added the operator-> to annotated_arg

Added e2e test

### CODE REVIEW

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc [hunk 1]
```diff
@@ -565,6 +565,7 @@ class annotated_ref {
     annotated_ref(T*);
     annotated_ref(const annotated_ref&) = default;
     operator T() const;
+    operator T&();
```
Comment (tiwaria1) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc:568]: Don't we need to make this const too?

**Review by tiwaria1 — APPROVED** [review]
Please add some cpp test lines that invoke all these operators to verify functionality at just the cpp level - is that possible?

**Review by tiwaria1 — CHANGES_REQUESTED** [review]
Commented

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc (hunk 1 above)
Comment (broxigarchen) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc:579]: We should not as we need this to be a modifiable LValue

**Comment by broxigarchen** [comment]
> Please add some cpp test lines that invoke all these operators to verify functionality at just the cpp level - is that possible?

added e2e test

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc [hunk 2]
```diff
@@ -565,6 +576,8 @@ class annotated_ref {
...
```
Comment (rolandschulz) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc:579]: We should not return a non-annotated reference. This break the whole reason why we have annotated-ref (to make sure annotations don't get silently dropped).

**Review by rolandschulz — CHANGES_REQUESTED** [review]

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc (hunk 1 above)
Comment (broxigarchen) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc]: removed this overload. Discussed with Roland, this approach seems error prone

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc (hunk 2 above)
Comment (broxigarchen) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc]: removed

**Comment by broxigarchen** [comment]
@rolandschulz @sergey-semenov  can you please help to review this again? Thanks!

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc [hunk 3]
```diff
@@ -210,6 +210,9 @@ class annotated_arg {
     operator T() noexcept;
     operator T() const noexcept;
 
+    // Available if type T is pointer type
+    T operator->() const noexcept;
```
Comment (rolandschulz) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:214]: This doesn't match the implementation. There it is `T*` return type. Which one is intended?

File: sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp [hunk 4]
```diff
@@ -151,17 +155,58 @@ class annotated_ref<T, detail::properties_t<Props...>> {
...
-  PROPAGATE_OP(-=)
-  PROPAGATE_OP(*=)
-  PROPAGATE_OP(/=)
-  PROPAGATE_OP(%=)
-  PROPAGATE_OP(^=)
-  PROPAGATE_OP(&=)
-  PROPAGATE_OP(|=)
+  this_t &operator++() {
+    T &t = annotated_load();
+    ++t;
```
Comment (rolandschulz) [inline sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp:160]: Shouldn't the store be annotated, too?

File: sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp [hunk 5]
```diff
@@ -151,17 +155,58 @@ class annotated_ref<T, detail::properties_t<Props...>> {
...
+    return *this;
+  }
+
+  this_t &operator++(int) {
+    T &t = annotated_load();
+    t++;
+    return *this;
```
Comment (rolandschulz) [inline sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp:167]: The post-increment operator should return the (reference to the) original value not the updated value.

File: sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp [hunk 6]
```diff
@@ -151,17 +155,58 @@ class annotated_ref<T, detail::properties_t<Props...>> {
...
 
-  PROPAGATE_OP(+=)
-  PROPAGATE_OP(-=)
-  PROPAGATE_OP(*=)
-  PROPAGATE_OP(/=)
-  PROPAGATE_OP(%=)
-  PROPAGATE_OP(^=)
-  PROPAGATE_OP(&=)
-  PROPAGATE_OP(|=)
+  this_t &operator++() {
```
Comment (rolandschulz) [inline sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp:158]: why return a ref to an annotated_ref?

File: sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp [hunk 7]
```diff
@@ -151,17 +155,58 @@ class annotated_ref<T, detail::properties_t<Props...>> {
...
+    t--;
+    return *this;
+  }
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define PROPAGATE_BINARY_OP(op)                                                \
+  annotated_ref operator op##=(const T &rhs) {                                 \
+    (*__builtin_intel_sycl_ptr_annotation(                                     \
+        m_Ptr, detail::PropertyMetaInfo<Props>::name...,                       \
+        detail::PropertyMetaInfo<Props>::value...)) op## = rhs;                \
```
Comment (rolandschulz) [inline sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp:187]: Shouldn't we annotate both the load and store?

Even if this isn't needed, I would like us to specify and implement the compound operators using the non-compound operators so it is possible to implement on top of builtin load+store. We agreed in https://github.com/intel/llvm/pull/10124 that this is the correct solution.

BTW: You didn't do the equivalent update of the spec for this changes.

**Review by rolandschulz — CHANGES_REQUESTED** [review]

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc (hunk 3 above)
Comment (broxigarchen) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:214]: This `T` in the docs should match the whole type, while in the implementation there are two subclass. That `T` is after removing the *

**Review by rolandschulz — APPROVED** [review]
LGTM

**Review by GarveyJoe — APPROVED** [review]

**Review by tiwaria1 — APPROVED** [review]

**Pull request merged**